### PR TITLE
KEP-6030: Dynamic resize of memory-backed volumes beta update

### PR DESCRIPTION
**Pull request by natasha41575**
Dynamic resize of memory-backed volumes beta update

Issue: https://github.com/kubernetes/enhancements/issues/6030

### CODE REVIEW

**Comment by natasha41575** [comment]
/assign @tallclair 
node

/assign @jsafrane 
storage

/assign @msau42 
storage and API

/assign @kannon92 
PRR

File: keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md [hunk 1]
```diff
@@ -568,20 +595,20 @@ Below are some examples to consider, in addition to the aforementioned [maturity
...
+- [X] Unit test coverage is complete.
+- [X] E2E tests implemented in CI verify successful upsize and downsize operations in a live cluster.
 
 #### Beta
-- Gather feedback from users.
-- Metrics are defined and implemented.
-- Kubelet emits a warning event when `sizeLimit` exceeds pod-level memory limits.
-- The plan evaluates whether integration tests are required.
-- Edge cases around "shrinkage safety" are resolved (e.g., reporting distinct errors when attempting to shrink a volume below current usage).
+- [ ] Gather feedback from users.
```
Comment (kannon92) [inline keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md:603]: How are you planning to gather feedback for this feature?

File: keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md
Comment (kannon92) [inline keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md:684]: This isn't quite true right.

`pkg/kubelet/eviction/eviction_manager.go:550` disables emptyDirLimitEviction for all medium: Memory volumes whenever the gate is on — not just for pods using resize:

```
if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScalingMemoryBackedVolumes) && source.EmptyDir.Medium == v1.StorageMediumMemory {
    continue
}
```

At beta default-on this silently changes behavior for every existing workload with a sized memory-backed emptyDir: they get ENOSPC on write instead of pod eviction + reschedule.

File: keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md [hunk 2]
```diff
@@ -465,24 +448,24 @@ On startup, the container runtime helps to bind the Kubelet's host path into the
 In the existing `InPlacePodVerticalScaling` feature, a best-effort safety logic ensures that container memory limits are not decreased below usage.
 
 #### Alpha
-For Alpha, shrinkage safety checks for memory-backed volumes are out of scope. If a user attempts to shrink a volume below its current usage, the `tmpfs` remount fails at the kernel level with an `EINVAL` error. This error is propagated back to the Kubelet and surfaced in the `PodResizeInProgress` condition as `mount point not mounted or bad option`. The Kubelet will periodically retry the resize operation until it succeeds, is cancelled, or the Pod terminates.
+For Alpha, shrinkage safety checks for memory-backed volumes are out of scope. If a user attempts to shrink a volume below its current usage, `/bin/mount` fails at the kernel level (e.g. on modern Linux kernels using the `fsconfig` tmpfs mount API, returning `fsconfig() failed: tmpfs: Too small a size for current use.` with exit status 32). This error is propagated back to the Kubelet and surfaced in the `PodResizeInProgress` condition. The Kubelet will periodically retry the resize operation until it succeeds, is cancelled, or the Pod terminates.
```
Comment (kannon92) [inline keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md:451]: Do we have any ideas how this will behave on windows? Or what version of the linux kernel where this behavior is not expected?

File: keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/kep.yaml [hunk 3]
```diff
@@ -50,5 +50,6 @@ disable-supported: true
 
 # The following PRR answers are required at beta release
 metrics:
-  - TBD
+  - volume_requested_resizes_total
```
Comment (kannon92) [inline keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/kep.yaml:53]: These metric names are not correct. I think you'd need kubelet in the name.

File: keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md [hunk 4]
```diff
@@ -966,6 +1017,10 @@ details). For now, we leave it here.
 
 ###### How does this feature react if the API server and/or etcd is unavailable?
 
+If the API server or etcd is unavailable, users will not be able to submit new resize requests. Existing requests will be processed by the Kubelet, but the pod status will not be updated with the resize status. 
+
+Once the API server or etcd is available again, new requests will be accepted and processed normally.
+
 ###### What are other known failure modes?
```
Comment (kannon92) [inline keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md:1024]: AI is calling out cgroup v1 as a potentail failure case.

8. Cgroup v1 rejection is invisible until it fails. NDF discovery is gate-only (Discover just checks FeatureGates.Enabled), so on a cgroup v1 node the API server accepts the resize and the kubelet then parks it as Infeasible. With the gate default-on at beta this is a new stuck state with no pre-flight signal. Add it to the known failure modes with detection (kubelet_pod_infeasible_resizes_total) and mitigation, or make Discover require cgroup v2.

File: keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md
Comment (kannon92) [inline keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md:750]: Aren't you disabling evictions in place of kernel OOM enforcement? Why would evictions increase?

Comment (kannon92) [inline keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md:869]: Should you mention NodeDeclaredFeatures?

File: keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md [hunk 5]
```diff
@@ -966,6 +1017,10 @@ details). For now, we leave it here.
 
 ###### How does this feature react if the API server and/or etcd is unavailable?
 
+If the API server or etcd is unavailable, users will not be able to submit new resize requests. Existing requests will be processed by the Kubelet, but the pod status will not be updated with the resize status.
```
Comment (kannon92) [inline keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md:1020]: Reading this I was thinking of static pods.

Maybe you could mention that this feature is also disabled for static pods?

File: keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md (hunk 1 above)
Comment (natasha41575) [reply keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md:603]: The feature was requested by Anyscale (for the IPPR Ray autoscaler integration); we are in contact them and will be engaging them for feedback.

We also have a few other customers internally that we have plans to engage with for feedback. So I anticipate that we will have enough here before code freeze in ~2 months.

File: keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md [hunk 6]
```diff
@@ -394,13 +381,15 @@ Init containers are commonly used to pre-populate a memory-backed volume (e.g.,
...
-*   The Kubelet does not block the resize but issues a **Warning Event**.
-*   If `sizeLimit` > cgroup memory limits, the volume's `ENOSPC` protection is effectively disabled, as the container is OOM-killed before the volume quota is reached. This is consistent with what pod creation permits today (see [Restricting sizeLimit to not exceed memory limits](#restricting-sizelimit-to-not-exceed-memory-limits)).
+When mounting or resizing a memory-backed volume, the Kubelet today computes the actual mounted `tmpfs` capacity as:
+$$\text{capacity} = \min(\text{NodeAllocatableMemory}, \text{PodMemoryLimit}, \text{VolumeSizeLimit})$$
+
+* If `sizeLimit` is omitted (`nil` or `0`), the volume size defaults to the total pod memory limit (or node allocatable if pod limits are unset).
+* If `sizeLimit` is specified and exceeds the total pod memory limit (e.g. 256Mi on a 200Mi pod), the mounted `tmpfs` volume capacity is capped to the total pod memory limit (200Mi). The Kubelet emits a `Warning VolumeSizeExceedsPodMemoryLimit` event during admission.
```
Comment (xing-yang) [inline keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md:392]: Do you need to update the OOM-kills section accordingly?

**Review by tallclair — COMMENTED** [review]
Approved for sig-node
/assign @dchen1107 
For official approval.

File: keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md
Comment (natasha41575) [reply keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md]: Somewhat. Pods today (before this feature) already get ENOSPC on write when workloads attempt to surpass the `sizeLimit` for memory-backed emptyDir volumes. The section in this KEP that talks about this is intended to be more like background information / preexisting context.

So I don't think the eviction manager for memory volumes was ever effective in practice, since the limit was hard-enforced by the kernel. I put it behind the feature gate as a safety measure, but I think the failure mode in question does not occur under normal operations. 

That said, disabling eviction mgr for memory-backed volumes is still a change to default behavior, so I'll update this section accordingly.

File: keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md (hunk 2 above)
Comment (natasha41575) [reply keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md]: The feature is disabled entirely on windows.

For linux, versions >= 5.2 use the new fsconfig mount API (`fsconfig() failed: tmpfs: Too small a size for current use`), while older kernels (< 5.2) or older mount binaries fail with the legacy mount failure (`mount point not mounted or bad option` or `cannot remount tmpfs read-write, is write-protected`). In both cases, the kernel safely rejects the shrink below active memory without corrupting data, and Kubelet surfaces the mount error directly in PodResizeInProgress.

I added this information to the KEP.

File: keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/kep.yaml (hunk 3 above)
Comment (natasha41575) [reply keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/kep.yaml]: Yeah all the metrics will have the `kubelet_` prefix. Fixed!

File: keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md (hunk 4 above)
Comment (natasha41575) [reply keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md:1032]: Added, thanks!

File: keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md
Comment (natasha41575) [reply keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md]: > Aren't you disabling evictions in place of kernel OOM enforcement? 

Kernel OOM enforcement is already in place (before this feature), which means that evictions never occurred for memory-backed volumes in practice. We're disabling these evictions, but related to my other comment https://github.com/kubernetes/enhancements/pull/6320#discussion_r4057725630, I don't think this is a meaningful change

But you're right, this section is wrong, there's no reason for evictions to increase, took that out.

Comment (natasha41575) [reply keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md:875]: I thought this question was about whether this feature depends on other out-of-tree (i.e. non-k8s) services running in the cluster, rather than other FGs? I added NodeDeclaredFeatures here though

File: keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md (hunk 5 above)
Comment (natasha41575) [reply keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md:1026]: done, thanks!

File: keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md (hunk 6 above)
Comment (natasha41575) [reply keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md:393]: updated the oom-kills section with this information!

**Review by kannon92 — COMMENTED** [review]
/approve
for PRR

**Comment by natasha41575** [comment]
/hold

just to gate the merge

File: keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md [hunk 7]
```diff
@@ -802,18 +842,20 @@ These goals will help you determine what you need to measure (SLIs) in the next
 question.
 -->
 
+- Resize requests should succeed (`kubelet_pod_resize_duration_milliseconds{"error=true"}` should be low)
```
Comment (xing-yang) [inline keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md:845]: `kubelet_pod_resize_duration_milliseconds{"error=true"}` -> 
`kubelet_pod_resize_duration_milliseconds{error="true"}`

File: keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md [hunk 8]
```diff
@@ -238,58 +243,42 @@ While the controller can use `InPlacePodVerticalScaling` to scale up a container
...
     *   *Admission during Resize*: Upon admitting a resize update, the `AllocationManager` updates the Kubelet's local checkpoint with the newly acknowledged capacity.
     *   *Scope*: Maintained only for memory-backed `emptyDir` volumes.
-*   **Actual State**: Represented by the new `VolumeStatus.EmptyDir.SizeLimit`.
-    *   *Definition*: Represents the actual mounted filesystem capacity reported by the mount/remount call.
-    *   *Pod Creation*: Remains unpopulated (`nil`) until Kubelet completes the initial volume mount setup and receives capacity confirmation from the mounter.
-    *   *Pod Resize*: During an active resize operation, this field retains the *old* mounted size until the Kubelet successfully actuates the `tmpfs` remount, at which point it is updated to the new actual capacity.
-    *   *Default/Fallback*: Remains unpopulated (`nil`) for all non-memory-backed volumes.
-
-The new status field `Pod.Status.ContainerStatuses[].VolumeMounts[].VolumeStatus.EmptyDir.SizeLimit` is only populated for memory-backed volumes, and remains unpopulated for other types of volumes. The `VolumeStatus` Go struct is updated as follows:
```
Comment (xing-yang) [inline keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md:267]: This is still mentioned in the Rollback and Rollout sections.

File: keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md [hunk 9]
```diff
@@ -777,13 +817,13 @@ and operation of this feature.
...
-- [ ] Events
-  - Event Reason: 
-- [ ] API .status
-  - Condition name: 
-  - Other field: 
-- [ ] Other (treat as last resort)
-  - Details:
+- [X] Events
+  - Event Reason: `ResizeStarted` (emitted when kubelet accepts a resize request, along with the new requested size)
+  - Event Reason: `ResizeCompleted` (emitted when kubelet successfully resizes the volume)
```
Comment (xing-yang) [inline keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md:822]: What about ResizeError?

File: keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md [hunk 10]
```diff
@@ -994,8 +1044,24 @@ For each of them, fill in the following information by copying the below templat
...
 -->
 
+- **Volume shrinkage below current in-use memory**:
+  - **Detection**: The `PodResizeInProgress` condition in pod status is set to `Status: True`, `Reason: Error` with a message from the mount failure (e.g., `tmpfs: Too small a size for current use.`), and a `Warning ResizeError` event is emitted on the pod.
+  - **Mitigations**: Running containers are unaffected (0 container restarts) and the volume remains mounted at its previous capacity. The workload can delete files to reduce usage, or the user/autoscaler can update the pod spec with a higher `sizeLimit` or cancel the resize.
+  - **Diagnostics**: Kubelet logs at `v=2` with `Resizing emptyDir volume` and mount error details.
+  - **Testing**: Covered by unit tests in `pkg/volume/emptydir` and node E2E tests verifying retry behavior and condition reporting on shrink failures.
+
+- **Resize requested on a node running cgroups v1**:
+  - **Detection**: The resize request is marked `Infeasible` in Pod status (`PodResizePending: Infeasible`), the metric `kubelet_pod_infeasible_resizes_total` is incremented, and a `Warning ResizeInfeasible` event is emitted on the Pod.
```
Comment (xing-yang) [inline keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md:1054]: Can you add `kubelet_pod_infeasible_resizes_total` to the metrics section?

Comment (natasha41575) [reply keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md:1066]: added `kubelet_pod_infeasible_resizes_total` to all metrics sections, thanks!

File: keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md (hunk 9 above)
Comment (natasha41575) [reply keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md:830]: SG, I added `ResizeError` and `ResizeInfeasible` to the list here for completeness

File: keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md (hunk 8 above)
Comment (natasha41575) [reply keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md:267]: Updated this! Thanks for the catch.

File: keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md (hunk 7 above)
Comment (natasha41575) [reply keps/sig-node/6030-dynamic-resize-of-memory-backed-volumes/README.md]: fixed, thanks!

**Comment by xing-yang** [comment]
lgtm

**Comment by dchen1107** [comment]
cc/ @andrewsykim etc. from OSS Ray community.
